src/gpb_parse.yrl

```diff
@@ -1059,60 +1059,95 @@ opt_tuple_to_atom_if_defined_true(Opt, Opts) ->
-prefix_suffix_msgs(Prefix, Suffix, ToLower, Defs) ->
+prefix_suffix_msgs(Prefix, Suffix, ToLowerOrSnake, Defs) ->
     lists:map(fun({{msg,Name}, Fields}) ->
-                      {{msg,prefix_suffix_name(Prefix, Suffix, ToLower, Name)},
-                       prefix_suffix_fields(Prefix, Suffix, ToLower, Fields)};
+                      {{msg,prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name)},
+                       prefix_suffix_fields(Prefix, Suffix, ToLowerOrSnake, Fields)};
                  ({{extensions,Name}, Exts}) ->
                       {{extensions,
-                        prefix_suffix_name(Prefix, Suffix, ToLower, Name)},
+                        prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name)},
                        Exts};
                  ({{service,Name}, RPCs}) ->
-                      {{service,maybe_tolower_name(Name, ToLower)},
-                       prefix_suffix_rpcs(Prefix, Suffix, ToLower, RPCs)};
+                      {{service, maybe_tolower_or_snake_name(Name, ToLowerOrSnake)},
+                       prefix_suffix_rpcs(Prefix, Suffix, ToLowerOrSnake, RPCs)};
                  ({package,Name}) ->
-                      {package,maybe_tolower_name(Name,ToLower)};
+                      {package, maybe_tolower_or_snake_name(Name,ToLowerOrSnake)};
                  ({proto3_msgs,Names}) ->
                       {proto3_msgs,
-                       [prefix_suffix_name(Prefix, Suffix, ToLower, Name)
+                       [prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name)
                         || Name <- Names]};
                  (OtherElem) ->
                       OtherElem
               end,
               Defs).
 
-prefix_suffix_fields(Prefix, Suffix, ToLower, Fields) ->
+prefix_suffix_fields(Prefix, Suffix, ToLowerOrSnake, Fields) ->
     lists:map(
       fun(#?gpb_field{type={msg,MsgName}}=F) ->
-              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLower, MsgName),
+              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, MsgName),
               F#?gpb_field{type={msg,NewMsgName}};
+         (#?gpb_field{type={map,KeyType,{msg,MsgName}}}=F) ->
+              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, MsgName),
+              F#?gpb_field{type={map,KeyType,{msg,NewMsgName}}};
+         (#gpb_oneof{fields=Fs}=F) ->
+              Fs2 = prefix_suffix_fields(Prefix, Suffix, ToLowerOrSnake, Fs),
+              F#gpb_oneof{fields=Fs2};
          (#?gpb_field{}=F) ->
               F
       end,
       Fields).
 
-prefix_suffix_name(Prefix, Suffix, ToLower, Name) ->
-    Name1 = maybe_tolower_name(Name, ToLower),
-    Name2 = lists:concat([Prefix, Name1, Suffix]),
+prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name) ->
+    Name1 = maybe_tolower_or_snake_name(Name, ToLowerOrSnake),
+    Prefix1 = case Prefix of
+                  {by_proto, PrefixList} ->
+                      proplists:get_value(Name1, PrefixList, "");
```

Comment:
hmm... just remembered I hadn't taken a look at this part. Sorry for doing it late. And also sorry in advance for bringing up a naming issue. :)

The prefix naming adds the option to specify prefix for individual messages on a per-message basis. Earlier, we discussed prefixing on a per-proto-file basis. I'm fine with either way, I think, but now that it is on a per-message basis, I see two things:
1. The `by_proto` name should be something else. `by_msg` comes to mind, but I'm of course open to other alternatives
2. The key that one should specify for prefixing messages is the name *after* any conversions (to-lower/snake etc) but since this is far from obvious, I think the `by_msg` (or whatever) sub-option would deserve a line or two in the (oh-so-long) doc text for `gpb_compile:file`.